### PR TITLE
Use testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe to avoid interference with other tests running in parallel

### LINKED ISSUE #3481: Intermittent test failure in `BasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe`
Intermittent test failure in `BasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe`

https://builds.apache.org/job/pulsar_precommit_cpp/5598/console

```
[93/145] BasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe (236 ms)
[0;33mNote: Google Test filter = BasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe
[m[0;32m[==========] [mRunning 1 test from 1 test case.
[0;32m[----------] [mGlobal test environment set-up.
[0;32m[----------] [m1 test from BasicEndToEndTest
[0;32m[ RUN      ] [mBasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe
2019-01-30 03:37:01.986 INFO  Client:88 | Subscribing on Topic :persistent://public/default/test-prevent-dup-consumers
2019-01-30 03:37:01.988 INFO  ConnectionPool:63 | Created connection for pulsar://localhost:6650
2019-01-30 03:37:01.989 INFO  ClientConnection:293 | [127.0.0.1:60310 -> 127.0.0.1:6650] Connected to broker
2019-01-30 03:37:01.996 INFO  HandlerBase:53 | [persistent://public/default/test-prevent-dup-consumers, my-only-sub, 0] Getting connection from pool
2019-01-30 03:37:02.179 INFO  ConsumerImpl:170 | [persistent://public/default/test-prevent-dup-consumers, my-only-sub, 0] Created consumer on broker [127.0.0.1:60310 -> 127.0.0.1:6650]
2019-01-30 03:37:02.179 INFO  Client:88 | Subscribing on Topic :persistent://public/default/test-prevent-dup-consumers
2019-01-30 03:37:02.180 INFO  ConsumerImpl:222 | [persistent://public/default/test-prevent-dup-consumers, my-only-sub, 0] Unsubscribing
2019-01-30 03:37:02.192 WARN  ClientConnection:870 | [127.0.0.1:60310 -> 127.0.0.1:6650] Received error response from server: 10 -- req_id: 1
2019-01-30 03:37:02.193 WARN  ConsumerImpl:257 | [persistent://public/default/test-prevent-dup-consumers, my-only-sub, 0] Failed to unsubscribe: BrokerMetadataError
/pulsar/pulsar-client-cpp/tests/BasicEndToEndTest.cc:2783: Failure
Value of: consumerA.unsubscribe()
  Actual: BrokerMetadataError
Expected: ResultOk
Which is: Ok
2019-01-30 03:37:02.193 WARN  ConsumerImpl:98 | [persistent://public/default/test-prevent-dup-consumers, my-only-sub, 0] Destroyed consumer which was not properly closed
2019-01-30 03:37:02.193 INFO  ConsumerImpl:813 | [persistent://public/default/test-prevent-dup-consumers, my-only-sub, 0] Closing consumer for topic persistent://public/default/test-prevent-dup-consumers
[0;31m[  FAILED  ] [mBasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe (208 ms)
[0;32m[----------] [m1 test from BasicEndToEndTest (208 ms total)

[0;32m[----------] [mGlobal test environment tear-down
[0;32m[==========] [m1 test from 1 test case ran. (208 ms total)
[0;32m[  PASSED  ] [m0 tests.
[0;31m[  FAILED  ] [m1 test, listed below:
[0;31m[  FAILED  ] [mBasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe

 1 FAILED TEST
[93/145] BasicEndToEndTest.testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe returned/aborted with exit code 1 (236 ms)
```

@lovelle Can you take a look at this one?

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

As reported in #3481, the `testDupConsumersOnSharedModeNotThrowsExcOnUnsubscribe` is failing intermmittently. 

The problem is due to fact that C++ tests are being executed in parallel and that this test was using the same topic / subscription name as another test. When these 2 tests ran at the same time, the test would fail.

### Modifications

 * Use unique name for test topic
 * Added info message when we're returning the shared consumer instance.

Fixes #3481

### CODE REVIEW

**Comment by lovelle** [comment]
Sorry about that! this is entirely my fault :disappointed:
Excellent using c++ auto in for loop!

**Comment by merlimat** [comment]
No worries, the code and the test were all correct. I had forgot that we run the tests in paralle: :)

File: pulsar-client-cpp/lib/ClientImpl.cc [hunk 1]
```diff
@@ -342,10 +342,11 @@ void ClientImpl::subscribeAsync(const std::string& topic, const std::string& con
             return;
         } else if (conf.getConsumerType() == ConsumerShared) {
             ConsumersList consumers(consumers_);
-            for (ConsumersList::iterator it = consumers.begin(); it != consumers.end(); ++it) {
-                ConsumerImplBasePtr consumer = it->lock();
+            for (auto& weakPtr : consumers) {
+                ConsumerImplBasePtr consumer = weakPtr.lock();
                 if (consumer && consumer->getSubscriptionName() == consumerName && !consumer->isClosed()) {
                     lock.unlock();
+                    LOG_INFO("Reusing existing consumer instance for " << topic << " -- " << consumerName);
```
Comment (lovelle) [inline pulsar-client-cpp/lib/ClientImpl.cc]: Maybe a `warn` would be better to alert the user of a miss use?

Comment (ivankelly) [reply pulsar-client-cpp/lib/ClientImpl.cc]: I would suggest to go the other way if anything. This isn't an something the user needs to be concerned about, so maybe it should be debug.

Comment (merlimat) [reply pulsar-client-cpp/lib/ClientImpl.cc]: The reason I added the log is that when looking at the logs one can now see `Subscribing.. ` but there's no corresponding `Subscribed... ` message. I think info level should be ok to keep the consistency with the rest of log messages.

Comment (lovelle) [reply pulsar-client-cpp/lib/ClientImpl.cc]: Great :+1:

**Comment by merlimat** [comment]
run integration tests

**Comment by merlimat** [comment]
run integration tests

**Comment by sijie** [comment]
run integration tests

**Comment by sijie** [comment]
run integration tests

**Pull request merged**